### ISSUE TITLE
Dynamically imported client component with variable name returns error in dev but works fine in prod

### ISSUE THREAD

**Issue by songhobby** [vercel/next.js#44946]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.2.0: Fri Nov 11 02:08:47 PST 2022; root:xnu-8792.61.2~4/RELEASE_X86_64
    Binaries:
      Node: 16.17.0
      npm: 8.15.0
      Yarn: 1.22.19
      pnpm: N/A
    Relevant packages:
      next: 13.1.2
      eslint-config-next: 13.1.2
      react: 18.2.0
      react-dom: 18.2.0

### Which area(s) of Next.js are affected? (leave empty if unsure)

App directory (appDir: true), Dynamic imports (next/dynamic)

### Link to the code that reproduces this issue

https://stackblitz.com/edit/vercel-next-js-zxfkvg?file=app/page.tsx

### To Reproduce

yarn dev

### Describe the Bug

When dynamically importing a client component with variable name which uses React context, next returns error on dev but works fine on prod. Note: dynamic import with static name works in both environment.

### Expected Behavior

`yarn dev` errors and the error says `TypeError: Cannot read properties of null (reading 'useContext')`
`yarn build && yarn start` works fine.

### Which browser are you using? (if relevant)

Chrome

### How are you deploying your application? (if relevant)

Vercel

**Comment by kachkaev** [comment]
Duplicate of https://github.com/vercel/next.js/issues/44946?

**Comment by Fredkiss3** [comment]
Sorry if my question is out of context, but you can use `next/dynamic` for this right ?

**Comment by songhobby** [comment]
> Sorry if my question is out of context, but you can use `next/dynamic` for this right ?

We need a programmatic way to load a component of variable name

**Comment by songhobby** [comment]
> Duplicate of / Related to #44083?

I'm not sure if they are caused by the same problem

**Comment by huozhi** [comment]
confirmed it's the same issue as #49382 , will close this for now and track there

**Comment by Tineshnehete** [comment]
Hi there ,
I also got the same issue with dynamic import and i tried alter methods also.
With ```dynamic(()=>import())``` method if you use the path which have any variable dependency  like shown in following snippet it leads to unexpected behavior like module not found or ```TypeError: Cannot read properties of null (reading 'useContext')``` .
```
import(`${path}`) // OR
import(path) // OR
import('strpath' + path)
```
And the same problem with ```require()``` , ```React.lazy()```.

but instead of path depending on var and you add explicite path like ``` import('full/path/with/no/var') ``` it works fine.

So for until it fixes we can use a dict to define views and use theme .
```
const Views = {
comp1: dynamic(()=>import("path/to/comp1")),
comp2: dynamic(()=>import("path/to/comp2")),
}
const MyPage = ({params})=>{
    const Comp = Views[params.view_name];
    return <><Comp /></>
}
```

**Comment by huozhi** [comment]
This problem is fixed in `next@13.4.5-canary.9`
